app/models/test_group_result.rb

```diff
@@ -0,0 +1,73 @@
+class TestGroupResult < ApplicationRecord
+  has_many :test_results, dependent: :destroy
+  belongs_to :test_group
+  belongs_to :test_run
+
+  validates :marks_earned, :marks_total, presence: true, numericality: { greater_than_or_equal_to: 0 }
+  validates :time, presence: true, numericality: { greater_than_or_equal_to: 0, only_integer: true }
+
+  def create_test_result(name:, output: '', marks_earned: 0.0, marks_total: 0.0, status: 'error', time: nil)
```

Comment:
I actually think it's better to remove this method altogether and stick with the Rails-conventional `test_results.create(...)`. You can pick default values for the attributes you want, but otherwise I'd just require passing in things like `status` explicitly.